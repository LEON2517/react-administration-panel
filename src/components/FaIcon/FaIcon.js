import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';


/**
  * Сахар для  font-awesome
  *
  * @param {Object} props - props
  * @property {boolean} spin -  fa-spin
  * @property {string} name -  название иконки
  * @property {string} className - название класса
  */
export default function FaIcon(props) {
  const nextClassNames = classnames({
    fa: true,
    [`fa-${props.name}`]: true,
    'fa-spin': props.spin,
  }, props.className);
  return (<span className={nextClassNames} />);
}

FaIcon.propTypes = {
  spin: PropTypes.bool,
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
};

FaIcon.defaultProps = {
  spin: false,
  className: '',
};
