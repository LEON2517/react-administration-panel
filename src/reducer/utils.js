/**
 * @module reducer/utils
 */
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/dom/ajax';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/do';
import 'rxjs/add/operator/map';
import { deprecate } from 'common';


/**
 * Исключение для редисера
 *
 * @class ActionError
 * @param {string} message - сообщение об ошибке
 */
export function ActionError(message) {
  this.name = 'ActionError';
  this.message = message;
  this.stack = (new Error()).stack;
}
ActionError.prototype = new Error();

/**
 * создает редюсер для указаного типа.
 * Aвтоматически объединяет со state
 *
 * @param {string} type - тип для редисера
 * @param {callback} callback -  редюсер
 * @returns {Reducer} reducer
 * @example
 * createReducer('Foo event', () => { foo: 'bar'});
 */
export function createReducer(type, callback) {
  return (state, action) => {
    if (!action || !action.type) {
      throw new ActionError('Action should have type');
    }
    if (action.type === type) {
      return {
        ...state,
        ...callback(state, action),
      };
    }
    return state;
  };
}

/**
 * Cоздает  корневой редюсер из массива редюсеров
 *
 * @param {Array<Reducer>} reducers - массив функций редюсеров
 * @param {State} initialState -  первоначальное состояние для редьсера
 * @returns {Reducer} reducer
 */
export function mergeReducers(reducers, initialState) {
  return (state = initialState, action) => {
    if (reducers.length) {
      return reducers.reduce((nextState, callback) => callback(nextState, action), state);
    }
    return state;
  };
}

/**
 * сахар для action
 *
 * @param {string} type - тип для action
 * @returns {ActionCreator} function
 * @example
 * export const someAction = actionCreator('Some Event');
 * someAction({id: 222});
 * //returns { type: 'Some Event', id: 222}
 */
export function actionCreator(type) {
  return payload => ({ type, payload });
}

/**
 * deprecate action creater
 *
 * @param {Function} type - редюсер
 * @param {string} oldName - страрoе имя редюсера
 * @param {string} nextName - страрoе имя редюсера
 *
 * @returns {Function}
 */
export function deprecateActionCreator(type, oldName, nextName) {
  return deprecate(actionCreator(type), `${oldName} reducer actions`, nextName);
}


/**
 * Создает  epic  который делает запрос на backend
 *
 * @param {string} type - redux type
 * @param {string | function} urlMap - маппер для  url
 * @param {ActionCreator} nextAction - следующий action
 */
export function createApiEpic(type, urlMap, nextAction) {
  const nextUrlMap = typeof urlMap === 'string' ? () => urlMap : urlMap;
  const streamedRequest = action => Observable.ajax(nextUrlMap(action))
    .map(({ response }) => nextAction(response.result));

  return action$ => action$.ofType(type).mergeMap(streamedRequest);
}

