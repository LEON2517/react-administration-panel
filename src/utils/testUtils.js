/**
 * сахар для тестов
 * @module utils/testUtils
 */

import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/from';
import 'rxjs/add/observable/empty';
import { StyleSheetServer } from 'aphrodite';
import cssToObject from 'css-to-object';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

/**
 *  Проверка CSS для aphrodite
 *
 *  @param {ReactDOM} render - React дерево
 *  @param {callback} expection - Функкция для проверки css
 *  @returns {void} no
 *  @example no
 */
export function expectCSS(render, expection) {
  const { css } = StyleSheetServer.renderStatic(render);
  expection(cssToObject(css.content));
}

/**
 * Сахар - снапшот для ReactDOM
 *
 * @param {ReactDOM} reactDom - React дерево
 * @returns {ReactTree} tree -  отрендереное дерево
 * @example no
 */
export function expectSnapshot(reactDom) {
  const tree = renderer.create(reactDom);
  expect(tree).toMatchSnapshot();

  return tree;
}

/**
 * Виртуальный скролл
 *
 * @param {number} offset -  новая позиция скролла
 * @returns {void} no
 * @example no
 */
export function scrollTo(offset) {
  window.pageYOffset = offset;
  window.dispatchEvent(new window.UIEvent('scroll', { detail: offset }));
}

/**
 * Создает поток для тестов епиков
 *
 * @param {string} type - тип для потока
 * @returns {ObservableAction} action$
 * @example
 * const action$ = createActionStream(TYPES.REQUEST_EXCHANGES);
 * const stream$ = requestExchangesEpic(action$);
 */
export function createActionStream(type) {
  const action$ = Observable.from([{ type }]);
  action$.ofType = (expectedType) => {
    if (type === expectedType) {
      return action$;
    }
    return Observable.empty();
  };
  return action$;
}

/**
 * Мок для  Observable.ajax
 *
 * @param {Object} result - ответ ajax
 * @returns {MockObservalbe} no
 * @example
 * const exchanges = ['bittrex'];
 * mockAjax(exchanges);
 * // теперь любой вызов  Observable.ajax - вернет  ['bittrex']
 */
export function mockAjax(result) {
  Observable.ajax = jest.fn(() => Observable.from([{
    response: {
      result,
    },
  }]));
  return Observable.ajax;
}

/**
 * Тест для observable
 *
 * @param {Observable} stream$ -  тестируемый поток
 * @param {Callback} callback - проверка результатат
 * @returns {Promise} promise
 * @example
 * await expectObservable(
 *  stream$,
 *  action => expect(action).toEqual(loadExchangesAction(exchanges)),
 * );
 */
export function expectObservable(stream$, callback) {
  return new Promise((resolve, reject) => {
    stream$.subscribe((value) => {
      try {
        callback(value);
        resolve();
      } catch (error) {
        reject(error);
      }
    });
  });
}

/**
 * mock для express.response
 *
 * @param {Object} expected - Роут должен возвращать этот объект
 */
export function createTestedResponse(expected) {
  return {
    send(data) {
      expect(data).toEqual(expected);
    },
    json(data) {
      this.send(data);
    },
  };
}

/**
 * инстремент для тестов Express
 *
 * @class ExpressTester
 */
export class ExpressTester {
  /**
   * конструктор
   */
  constructor() {
    this.getRoutes = new Map();
    this.postRoutes = new Map();
    this.router = {
      get: (key, callback) => {
        this.getRoutes.set(key, callback);
      },
      post: (key, callback) => {
        this.postRoutes.set(key, callback);
      },
    };
  }
  /**
    * getter router
    *
    * @returns {MockRouter}
    */
  getRouter() {
    return this.router;
  }
  /**
    * Мок для  get запроса
    *
    * @param {string} url - адресс
    * @param {Object} expected - ожидаемый ответ
    * @param {MockRequest} request -  мок для express.request
    *
    */
  get(url, expected, request = null) {
    if (this.getRoutes.has(url)) {
      return this.getRoutes.get(url)(request, createTestedResponse(expected));
    }
    throw new Error(`GET url not existed: ${url}`);
  }
  /**
   * Мок для post запроса
   *
   * @param {string} url - адресс
   * @param {Object} expected - ожидаемый ответ
   * @param {MockRequest} request -  мок для express.request
   *
   */
  post(url, expected, request = null) {
    if (this.postRoutes.has(url)) {
      return this.postRoutes.get(url)(request, createTestedResponse(expected));
    }
    throw new Error(`POST url not existed: ${url}`);
  }
}


/**
 * Мок для редукс store
 *
 * @param {Object} initialState - началный state
 * @returns {ReduxStore} store
 */
export function mockStore(initialState) {
  const mockStoreCreator = configureStore([]);
  const store = mockStoreCreator(initialState);
  store.getRealState = store.getState;
  store.getState = () => {
    const state = store.getRealState();
    return new Proxy(state, {
      get(target, prop) {
        if (prop === 'searchMarket') {
          // eslint-disable-next-line no-console
          console.error('searchMarket reducers deprecated use marketsList');
        }
        return target[prop];
      },
    });
  };

  return store;
}

