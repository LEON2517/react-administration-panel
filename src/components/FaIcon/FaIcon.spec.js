import React from 'react';
import { expectSnapshot } from 'utils/testUtils';

import FaIcon from './FaIcon';

describe('FaIcon spec', () => {
  test('render name', () => {
    expectSnapshot(<FaIcon name="pensil" />);
  });
  test('render spin', () => {
    expectSnapshot(<FaIcon name="pensil" spin />);
  });
});
