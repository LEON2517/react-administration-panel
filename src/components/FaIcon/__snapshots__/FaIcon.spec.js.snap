// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FaIcon spec render name 1`] = `
<span
  className="fa fa-pensil"
/>
`;

exports[`FaIcon spec render spin 1`] = `
<span
  className="fa fa-pensil fa-spin"
/>
`;
